src/_path.h

```diff
@@ -939,79 +939,167 @@ void cleanup_path(PathIterator &path,
+               cubic curves */
+            if (code == CURVE3 && codes[code - 1][0] == '\0') {
+                quad2cubic(last_x, last_y, x[0], y[0], x[1], y[1], x, y);
+                code++;
+                size = 3;
+            }
 
+            if (!postfix) {
+                if ((p = __append_to_string(p, buffer, *buffersize, codes[code - 1])) == NULL) return 1;
+                if ((p = __append_to_string(p, buffer, *buffersize, " ")) == NULL) return 1;
+            }
+
+            for (int i = 0; i < size; ++i) {
 #if PY_VERSION_HEX >= 0x02070000
-        char *str;
-        str = PyOS_double_to_string(x, 'g', precision, 0, NULL);
-        p += snprintf(p, *buffersize - (p - buffer), "%s", str);
-        PyMem_Free(str);
-        *p++ = ' ';
-        str = PyOS_double_to_string(y, 'g', precision, 0, NULL);
-        p += snprintf(p, *buffersize - (p - buffer), "%s", str);
-        PyMem_Free(str);
+                char *str;
+                str = PyOS_double_to_string(x[i], 'g', precision, 0, NULL);
+                if ((p = __append_to_string(p, buffer, *buffersize, str)) == NULL) {
+                    PyMem_Free(str);
+                    return 1;
+                }
+                PyMem_Free(str);
+                if ((p = __append_to_string(p, buffer, *buffersize, " ")) == NULL) return 1;
+                str = PyOS_double_to_string(y[i], 'g', precision, 0, NULL);
+                if ((p = __append_to_string(p, buffer, *buffersize, str)) == NULL) {
+                    PyMem_Free(str);
+                    return 1;
+                }
+                PyMem_Free(str);
+                if ((p = __append_to_string(p, buffer, *buffersize, " ")) == NULL) return 1;
 #else
-        char str[64];
-        PyOS_ascii_formatd(str, 64, format, x);
-        p += snprintf(p, *buffersize - (p - buffer), "%s", str);
-        *p++ = ' ';
-        PyOS_ascii_formatd(str, 64, format, y);
-        p += snprintf(p, *buffersize - (p - buffer), "%s", str);
+                char str[64];
+                PyOS_ascii_formatd(str, 64, format, x[i]);
+                if ((p = __append_to_string(p, buffer, *buffersize, str)) == NULL) return 1;
+                p = __append_to_string(p, buffer, *buffersize, " ");
+                PyOS_ascii_formatd(str, 64, format, y[i]);
+                if ((p = __append_to_string(p, buffer, *buffersize, str)) == NULL) return 1;
+                if ((p = __append_to_string(p, buffer, *buffersize, " ")) == NULL) return 1;
 #endif
+            }
+
+            if (postfix) {
+                if ((p = __append_to_string(p, buffer, *buffersize, codes[code - 1])) == NULL) return 1;
+            }
+
+            last_x = x[size - 1];
+            last_y = y[size - 1];
+        } else {
+            // Unknown code value
+            return 2;
+        }
 
-        --wait;
+        if ((p = __append_to_string(p, buffer, *buffersize, "\n")) == NULL) return 1;
     }
 
     *p = '\0';
```

Comment:
This should use `__append_to_string` too?  I think this can be an issue if the path is exactly the length of the buffer when the '\n' is added.